### PR TITLE
Fix incorrect static casts on variables on 32-bit platforms

### LINKED ISSUE #9366: Test failure (api/cpp/sep_log_api) on 32-bit architectures
I'm seeing the api/cpp/sep_log_api test fail on 32-bit architectures (i386, armel, armhf).  This is on Debian Sid with cvc5 1.0.3.

```
3165/3185 Test #3225: api/cpp/reset_assertions .....................................................................   Passed    0.01 sec
          Start 3226: api/cpp/sep_log_api
3166/3185 Test #3226: api/cpp/sep_log_api ..........................................................................Subprocess aborted***Exception:   0.02 sec
terminate called after throwing an instance of 'cvc5::CVC5ApiException'
  what():  Invalid argument '0' for '*d_node', expected Term to be a 64-bit integer value when calling getInt64Value()

          Start 3227: api/cpp/smt2_compliance
3167/3185 Test #3227: api/cpp/smt2_compliance ......................................................................   Passed    0.01 sec
```

Any advice on how to investigate further would be appreciated.  CC @nafur since you fixed a similar issue with this test?

### PR DESCRIPTION
**Pull request by aytey**
Before this PR, there is code inside of `cvc5` for both CLN and GMP, which is guarded by `CVC5_NEED_INT64_T_OVERLOADS`, that can lead to a loss of precision, due to unconditionally cast types such as `int64_t` into `long` without checking if the values fit.

For GMP, there were places (e.g., `Rational(int64_t n, int64_t d)`) where larger types (e.g., `int64_t`) got truncated into shorter types (e.g., `long`). This PR refactors and centralises the code to convert an `(u)int64_t` to an `mpz`, and makes sure all values that do not fit inside of `(un)signed long` go through `std::to_string`. Construction of an `mpq` goes through the same code path, and uses the `mpq_class(mpz_class)` constructor to avoid duplicated code (this is because the input term is an integer, so starting with an `mpz` is valid).

For CLN, there were places (e.g., `Rational(int64_t n)`) where larger types (e.g., `int64_t`) were truncated into shorter types (e.g., `long`). These casts were unnecessary as CLN has the required overloads to immediately support `int64_t` without requiring a cast to a different type, so have been removed.

I have tested this on the same 32-bit platform that exhibits the error for #9366 and confirm it resolves the issue.

Closes #9366.

Signed-off-by: Andrew V. Teylu <andrew@tey.lu>

### CODE REVIEW

File: src/util/rational_gmp_imp.h [hunk 1]
```diff
@@ -95,11 +95,11 @@ class Rational
   Rational(unsigned long int n) : d_value(n, 1) { d_value.canonicalize(); }
 
 #ifdef CVC5_NEED_INT64_T_OVERLOADS
-  Rational(int64_t n) : d_value(static_cast<long>(n), 1)
+  Rational(int64_t n) : d_value(static_cast<signed long int>(n), 1)
```
Comment (mpreiner) [inline src/util/rational_gmp_imp.h:98]: Shouldn't these be long long, similar to the other changes?

File: src/util/rational_gmp_imp.h [hunk 2]
```diff
@@ -95,11 +95,11 @@ class Rational
...
   {
     d_value.canonicalize();
   }
-  Rational(uint64_t n) : d_value(static_cast<unsigned long>(n), 1)
+  Rational(uint64_t n) : d_value(static_cast<unsigned long int>(n), 1)
```
Comment (mpreiner) [inline src/util/rational_gmp_imp.h:102]: Ditto

File: src/util/rational_gmp_imp.h [hunk 3]
```diff
@@ -127,12 +127,12 @@ class Rational
 
 #ifdef CVC5_NEED_INT64_T_OVERLOADS
   Rational(int64_t n, int64_t d)
-      : d_value(static_cast<long>(n), static_cast<long>(d))
+      : d_value(static_cast<signed long int>(n), static_cast<signed long int>(d))
```
Comment (mpreiner) [inline src/util/rational_gmp_imp.h:130]: Ditto

File: src/util/rational_gmp_imp.h [hunk 4]
```diff
@@ -127,12 +127,12 @@ class Rational
...
   {
     d_value.canonicalize();
   }
   Rational(uint64_t n, uint64_t d)
-      : d_value(static_cast<unsigned long>(n), static_cast<unsigned long>(d))
+      : d_value(static_cast<unsigned long int>(n), static_cast<unsigned long int>(d))
```
Comment (mpreiner) [inline src/util/rational_gmp_imp.h:135]: Ditto

File: src/util/rational_gmp_imp.h (hunk 1 above)
Comment (aytey) [reply src/util/rational_gmp_imp.h:98]: They got changed in the second commit ...

* GMP works on both Windows and Linux; and GMP *does not* have a `long long` overload (at least on Windows)
* CLN only works on Linux, and *does* have a `long long` overload

In actually gets a bit worse than this: if you make both CLN and GMP be `long int` then the separation logic API tests still fail on 32-bit Debian -- it seems it really has to be `(un)signed long long` to make CLN be happy on 32-bit Debian.

Basically, yes, I'd like them both to be `long long`, but `(un)signed long int` was the best I could get for GMP on Windows.

Comment (mpreiner) [reply src/util/rational_gmp_imp.h:98]: Alright, got it. Can you document this in the header accordingly? (Just to avoid that later changes try to "fix" this).

Comment (aytey) [reply src/util/rational_gmp_imp.h]: @mpreiner: okay, I implemented something, but something you might "hate": take a look at the most recent commit.

File: src/util/integer_gmp_imp.cpp [hunk 5]
```diff
@@ -54,10 +54,10 @@ Integer::Integer(int64_t z)
 }
 Integer::Integer(uint64_t z)
 {
-  if (std::numeric_limits<unsigned long int>::min() <= z
-      && z <= std::numeric_limits<unsigned long int>::max())
+  if (std::numeric_limits<integer_gmp_largest_unsigned>::min() <= z
+      && z <= std::numeric_limits<integer_gmp_largest_unsigned>::max())
   {
-    d_value = static_cast<unsigned long int>(z);
+    d_value = static_cast<rational_gmp_largest_unsigned>(z);
```
Comment (swt2c) [inline src/util/integer_gmp_imp.cpp:60]: Should this be `integer_gmp_largest_unsigned` here?

**Comment by aytey** [comment]
As a note about the "weirdness" with GMP:

* https://stackoverflow.com/a/75036967
* https://github.com/alisw/GMP/blob/master/gmpxx.h#L1514-L1524 (random mirror)
* https://gmplib.org/list-archives/gmp-discuss/2021-January/006622.html

Basically, GMP *does not* support `long long`, so we cannot use that. We then have to use the nearest, largest type, which is `(un)signed long int`.

Now, it is possible that if some part of cvc5 uses the GMP class like the CLN class, and passes in something larger than a 32-bit number, things will go awry.

I am not entirely sure I understand why the separation logic tests work on a 32-bit platform without CLN, given I would have expected the limit checks to overflow the GMP classes in the same way.

Still, this explains why we need to use `(un)signed long int` for GMP and `(un)signed long long` for CLN.

**Comment by aytey** [comment]
Ah! We should be using `mpz_import` in the GMP code; I’ll fix this over the weekend.

File: src/util/rational_gmp_imp.h (hunk 1 above)
Comment (aytey) [reply src/util/rational_gmp_imp.h]: The PR now takes a totally different approach, so these comments have been completely superseded.

**Comment by aytey** [comment]
> I am not entirely sure I understand why the separation logic tests work on a 32-bit platform without CLN, given I would have expected the limit checks to overflow the GMP classes in the same way.

This was because GMP checked if the input value fit inside of `(un)signed long`, and if it **did not**, it used `std::to_string`.

File: src/util/integer_cln_imp.h [hunk 6]
```diff
@@ -74,8 +74,8 @@ class Integer
   Integer(unsigned long int z) : d_value(z) {}
 
 #ifdef CVC5_NEED_INT64_T_OVERLOADS
-  Integer(int64_t z) : d_value(static_cast<long>(z)) {}
-  Integer(uint64_t z) : d_value(static_cast<unsigned long>(z)) {}
+  Integer(int64_t z) : d_value(z) {}
```
Comment (swt2c) [inline src/util/integer_cln_imp.h:77]: So CLN didn't need the casts to begin with?

Comment (aytey) [reply src/util/integer_cln_imp.h:77]: The short answer: yes

My guess is that whoever wrote the CLN implementation in CVC5, started by copying the GMP implementation -- when it came to the sections inside of `CVC5_NEED_INT64_T_OVERLOADS`, they didn't check if the casts were required or not (they're not). They likely didn't check because hitting that code is hard (only Windows or 32-bit Linux).

In summary: anywhere a term was casted to `long` in GMP was incorrect (now fixed) and anywhere a term was casted to `long` in CLN was unnecessary (now also fixed).

I had attempted to capture some of these points in the (updated) PR description; take a look and let me know if you feel I cover the intent of these changes.

Comment (aytey) [reply src/util/integer_cln_imp.h:77]: > My guess is that whoever wrote the CLN implementation in CVC5, started by copying the GMP implementation -- when it came to the sections inside of `CVC5_NEED_INT64_T_OVERLOADS`, they didn't check if the casts were required or not (they're not). They likely didn't check because hitting that code is hard (only Windows or 32-bit Linux).

I was wrong:

* https://github.com/cvc5/cvc5/commit/94b685e149119bbe75266481f6de3a0915e7c10a

I have no idea why those casts are there for CLN (maybe CLN didn't used to support `long long`?); maybe Morgan just made the code be consistent between GMP and CLN, without realising it wasn't required for CLN?

File: src/util/gmp_util.h [hunk 7]
```diff
@@ -33,6 +41,49 @@ inline size_t gmpz_hash(const mpz_t toHash) {
...
+   * construct result based on the value */
+  if (std::numeric_limits<gmp_type>::min() <= z
+      && z <= std::numeric_limits<gmp_type>::max())
+  {
+    result = static_cast<gmp_type>(z);
+  }
+  else
+  {
+    /* otherwise, use the string-based methods */
+    result = std::to_string(z);
```
Comment (mpreiner) [inline src/util/gmp_util.h:82]: I'm actually not sure how expensive the string conversion is, but since we are converting from [u]int64_t, we could also do the following, similar to what we do in Bitwuzla (https://github.com/bitwuzla/bitwuzla/blob/main/src/lib/bv/bitvector.cpp#L120):

for unsigned:
```
    uint32_t hi = static_cast<uint32_t>(z >> 32);
    uint32_t lo = static_cast<uint32_t>(z);
    mpz_init_set_ui(result, hi);
    mpz_mul_2exp(result, result, 32);
    mpz_add_ui(result, result, lo);
```

for signed:
```
    int32_t hi = static_cast<int32_t>(z >> 32);
    int32_t lo = static_cast<int32_t>(z);
    mpz_init_set_si(result, hi);
    mpz_mul_2exp(result, result, 32);
    mpz_add_ui(result, result, lo);
```

Comment (aytey) [reply src/util/gmp_util.h:82]: So, yes, that’s what I originally thought I’d be doing but using mpz_import (https://github.com/cvc5/cvc5/pull/10334#issuecomment-1936727239), but then I saw that cvc5 was already using the string-based method.

As a note, I didn’t write the string-based code, it was already in cvc5, I just moved it to make it usable across multiple classes.

Do you want me to move this code to be using mpz_import or similar, or leave it as-is, given cvc5 already uses the string-based approach in main?

Comment (mpreiner) [reply src/util/gmp_util.h:82]: Ok if it was already there, let's leave it as it is now.

Comment (aytey) [reply src/util/gmp_util.h:82]: Yep: https://github.com/cvc5/cvc5/blob/2576488d68ea43470f152ab4c6d7785b9f9fb51e/src/util/integer_gmp_imp.cpp#L52 (main at the time of writing).

Basically, everywhere that cvc5 goes from an `(u)int64` -> GMP-type needs to go through that check (or do something like you're suggesting).

I went for the simplest route of just making that functionality common, and then using it everywhere that GMP types are instantiated from `(u)int64`.

File: src/util/gmp_util.h [hunk 8]
```diff
@@ -20,6 +20,14 @@
 
 #include <gmpxx.h>
 
+/* the following includes are only required if CVC5_NEED_INT64_T_OVERLOADS is
+ * defined
+ * */
```
Comment (mpreiner) [inline src/util/gmp_util.h:25]:
```suggestion
 // The following includes are only required if CVC5_NEED_INT64_T_OVERLOADS is
 // defined.
```

File: src/util/rational_gmp_imp.h [hunk 9]
```diff
@@ -95,11 +95,10 @@ class Rational
   Rational(unsigned long int n) : d_value(n, 1) { d_value.canonicalize(); }
 
 #ifdef CVC5_NEED_INT64_T_OVERLOADS
-  Rational(int64_t n) : d_value(static_cast<long>(n), 1)
-  {
-    d_value.canonicalize();
-  }
-  Rational(uint64_t n) : d_value(static_cast<unsigned long>(n), 1)
+  /* to avoid truncation, we convert the input value to an mpz and then build
+   * the mpq */
```
Comment (mpreiner) [inline src/util/rational_gmp_imp.h:99]:
```suggestion
   // To avoid truncation, we convert the input value to an mpz and then build
   // the mpq.
```

File: src/util/rational_gmp_imp.h [hunk 10]
```diff
@@ -126,13 +125,13 @@ class Rational
   }
 
 #ifdef CVC5_NEED_INT64_T_OVERLOADS
-  Rational(int64_t n, int64_t d)
-      : d_value(static_cast<long>(n), static_cast<long>(d))
+  /* to avoid truncation, we convert the input value to an mpz and then build
+   * the mpq */
```
Comment (mpreiner) [inline src/util/rational_gmp_imp.h:129]:
```suggestion
   // To avoid truncation, we convert the input value to an mpz and then build
   // the mpq.
```

**Comment by mpreiner** [comment]
PR looks good to me, just some minor suggestions. @aytey can you squash the commits and sign it? Thanks!

File: src/util/rational_gmp_imp.h (hunk 10 above)
Comment (aytey) [reply src/util/rational_gmp_imp.h:129]: Done!

File: src/util/rational_gmp_imp.h (hunk 9 above)
Comment (aytey) [reply src/util/rational_gmp_imp.h]: Done!

File: src/util/gmp_util.h (hunk 8 above)
Comment (aytey) [reply src/util/gmp_util.h]: Done!

**Comment by aytey** [comment]
> PR looks good to me, just some minor suggestions. @aytey can you squash the commits and sign it? Thanks!

Done!

**Review by mpreiner — APPROVED** [review]
